src/java.base/macosx/native/libnio/fs/MacOSXWatchService.c

```diff
@@ -0,0 +1,237 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "jni.h"
+#include "jni_util.h"
+#include "nio_util.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <CoreServices/CoreServices.h>
+
+// MacOSXWatchService.callback()
+static jmethodID        callbackMID;
+// The instance of MacOSXWatchService that is associated with this thread
+static __thread jobject watchService;
+
+JNIEXPORT void JNICALL
+Java_sun_nio_fs_MacOSXWatchService_initIDs(JNIEnv* env, __unused jclass clazz)
+{
+    callbackMID = (*env)->GetMethodID(env, clazz,
+                                      "callback", "(J[Ljava/lang/String;J)V");
+}
+
+extern CFStringRef
+toCFString(JNIEnv *env, jstring javaString);
+
+static jboolean
+convertToJavaStringArray(JNIEnv* env,
+                         char **eventPaths,
+                         const jsize numEventsToReport,
+                         jobjectArray javaEventPathsArray)
+{
+    for (jsize i = 0; i < numEventsToReport; i++) {
+        const jstring path = JNU_NewStringPlatform(env, eventPaths[i]);
+        CHECK_NULL_RETURN(path, FALSE);
+        (*env)->SetObjectArrayElement(env, javaEventPathsArray, i, path);
+    }
+
+    return JNI_TRUE;
+}
+
+static void
+callJavaCallback(JNIEnv* env,
+                 jlong streamRef,
+                 jobjectArray javaEventPathsArray,
+                 jlong eventFlags)
+{
+    if (callbackMID != NULL && watchService != NULL) {
+        // We are called on the run loop thread, so it's OK to use
+        // the thread-local reference to the watch service.
+        (*env)->CallVoidMethod(env, watchService, callbackMID,
+                               streamRef, javaEventPathsArray, eventFlags);
+    }
+}
+
+/**
+ * Callback that is invoked on the run loop thread and informs of new
+ * file-system events from an FSEventStream.
+ */
+static void
+callback(__unused ConstFSEventStreamRef streamRef,
+         __unused void *clientCallBackInfo,
+         size_t numEventsTotal,
+         void *eventPaths,
+         const FSEventStreamEventFlags eventFlags[],
+         __unused const FSEventStreamEventId eventIds[])
+{
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
+    if (!env) { // Shouldn't happen as run loop starts from Java code
```

Comment:
Is this check, hence the `jvm` extern, really necessary?

Reply:
The check isn't strictly necessary, so I'm goign to drop it (there's a JNI version check in `nio_util.c`), but the `jvm` extern seems necessary because that's how `JNIEnv` is obtained and the latter is necessary to communicate the events back to the Java side.